lib/logstash/plugin_mixins/jdbc/statement_handler.rb

```diff
@@ -0,0 +1,126 @@
+# encoding: utf-8
+
+module LogStash module PluginMixins module Jdbc
+  class StatementHandler
+    def self.build_statement_handler(plugin, logger)
+      klass = plugin.use_prepared_statements ? PreparedStatementHandler: NormalStatementHandler
+      klass.new(plugin, logger)
+    end
+
+    attr_reader :statement, :parameters, :statement_logger
+
+    def initialize(plugin, statement_logger)
+      @statement = plugin.statement
+      @statement_logger = statement_logger
+      post_init(plugin)
+    end
+
+    def build_query(db, sql_last_value)
+      # override in subclass
+    end
+
+    def post_init(plugin)
+      # override in subclass, if needed
+    end
+  end
+
+  class NormalStatementHandler < StatementHandler
+    # Performs the query, respecting our pagination settings, yielding once per row of data
+    # @param db [Sequel::Database]
+    # @param sql_last_value [Integet|DateTime|Time]
+    # @yieldparam row [Hash{Symbol=>Object}]
+    def perform_query(db, sql_last_value)
+      query = build_query(db, sql_last_value)
+      if @jdbc_paging_enabled
+        query.each_page(@jdbc_page_size) do |paged_dataset|
+          paged_dataset.each do |row|
+            yield row
+          end
+        end
+      else
+        query.each do |row|
+          yield row
+        end
+      end
+    end
+
+    private
+
+    def build_query(db, sql_last_value)
+      parameters[:sql_last_value] = sql_last_value
+      query = db[statement, parameters]
+      statement_logger.log_statement_parameters(query, statement, parameters)
+      query
+    end
+
+    def post_init(plugin)
+      @parameter_keys = ["sql_last_value"] + plugin.parameters.keys
+      @parameters = plugin.parameters.inject({}) do |hash,(k,v)|
+        case v
+        when LogStash::Timestamp
+          hash[k.to_sym] = v.time
+        else
+          hash[k.to_sym] = v
+        end
+        hash
+      end
+    end
+  end
+
+  class PreparedStatementHandler < StatementHandler
+    attr_reader :name, :bind_values_array, :statement_prepared, :prepared
+
+    # Performs the query, ignoring our pagination settings, yielding once per row of data
+    # @param db [Sequel::Database]
+    # @param sql_last_value [Integet|DateTime|Time]
+    # @yieldparam row [Hash{Symbol=>Object}]
+    def perform_query(db, sql_last_value)
+      query = build_query(db, sql_last_value)
+      query.each do |row|
+        yield row
+      end
+    end
+
+    private
+
+    def build_query(db, sql_last_value)
+      # don't log statement count when using prepared statements for now...
```

Comment:
Does it make sense to log the statement, just not the count?

Reply:
In a round about way I guess you are correct. The decision of what to log is a concern of the CheckedCountLogger class with a hint from the flavour of statement handler. The PreparedStatementHandler should be able to unset the `@needs_check` ivar in some way - then the logging without count will take place.